### PR TITLE
First bunch of notebook tests.

### LINKED ISSUE #136: Provide notebook testing.
Add testing that will make sure
- a sequence of notebooks runs all their code cells without exception
- all links in the markdown cells are valid.

Make sure all notebooks throw an exception when the results are unexpected. For instance, a Transformers extension UDF may return an error message in the query result. This won't raise an exception and therefore won't be caught in testing. Need to assert that the error is empty explicitly.

### LINKED ISSUE #152: Urgent Bugfixes before Release
We found a few issues in the last review session.

- Local Model upload for transformers didn't work
- The link to the sagemaker init didn't work

### PR DESCRIPTION
**Pull request by ahsimb**
Fixes #152 
Fixes #136

### CODE REVIEW

File: test/notebooks/notebook_test_utils.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    Returns the temporary configuration store path and password.
+    """
+
+    # Create temporary secret store
+    store_path = tmp_path / 'tmp_config.sqlite'
+    # Should the password be constant or randomly generated is debatable.
+    # Strictly speaking, random password makes a test non-deterministic.
+    # On the other hand, it improves the security. If for some reason the
+    # temporary configuration store is not deleted, there will be no access
+    # to it since the password will be forgotten once the session is over.
```
Comment (tkilias) [inline test/notebooks/notebook_test_utils.py:106]: good question, in the CI and from the test setup, everything should be temporary, so nothing should survive for long. The main concerns could be the AWS credentials and other secrets. From that point, maybe it is actually a good idea, to use a random password.

I would also recommend to add finally block which removes the secret store file

File: test/notebooks/notebook_test_utils.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    """
+    A fixture for running a notebook.
+    """
+
+    store_path, store_password = access_to_temp_secret_store
+    return partial(run_notebook, store_file=str(store_path), store_password=store_password)
+
+
+@pytest.fixture
+def uploading_hack() -> Tuple[str, str]:
```
Comment (tkilias) [inline test/notebooks/notebook_test_utils.py:133]: ok know I understand why you need to inject a hack

Comment (tkilias) [reply test/notebooks/notebook_test_utils.py:133]: maybe mention, that  extracting the model in bucketfs takes a while, however the upload returns immediately

Comment (tkilias) [reply test/notebooks/notebook_test_utils.py:133]: we also document that with the hacks in the developer guide

**Review by tkilias — CHANGES_REQUESTED** [review]

**Comment by ahsimb** [comment]
Please ignore the Sagemaker test for now.

File: test/notebooks/nbtest_transformers.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    running_hack = (
+        'running_model',
+        textwrap.dedent("""
+            errors = %sql --with udf_output SELECT error_message FROM udf_output where error_message != 'None'
+            assert not errors, str(errors)
+        """)
+    )
+    running_pyx_hack = (
+        'running_model_pyx',
```
Comment (tkilias) [inline test/notebooks/nbtest_transformers.py:31]:
```suggestion
        'running_model_pyexasol',
```

reason, pyx sounds too much like cython files or some special python file extension.

File: test/notebooks/nbtest_transformers.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+def test_transformers(notebook_runner, uploading_hack, notebook_file) -> None:
+
+    running_hack = (
+        'running_model',
+        textwrap.dedent("""
+            errors = %sql --with udf_output SELECT error_message FROM udf_output where error_message != 'None'
+            assert not errors, str(errors)
+        """)
+    )
+    running_pyx_hack = (
```
Comment (tkilias) [inline test/notebooks/nbtest_transformers.py:30]:
```suggestion
    running_pyexasol_hack = (
```

File: test/notebooks/nbtest_transformers.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            reason='waiting for the Sacremoses tokenizer to be installed in the SLC')),
+        'zero_shot_classification.ipynb'
+    ]
+)
+def test_transformers(notebook_runner, uploading_hack, notebook_file) -> None:
+
+    running_hack = (
+        'running_model',
+        textwrap.dedent("""
+            errors = %sql --with udf_output SELECT error_message FROM udf_output where error_message != 'None'
```
Comment (tkilias) [inline test/notebooks/nbtest_transformers.py:26]:
```suggestion
            errors = %sql --with udf_output SELECT error_message FROM udf_output WHERE error_message != 'None'
```

File: test/notebooks/nbtest_sagemaker.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            Bucket=bucket_name,
+            CreateBucketConfiguration={
+                'LocationConstraint': os.environ["NBTEST_AWS_DEFAULT_REGION"]}
+        )
+    except s3_client.exceptions.BucketAlreadyOwnedByYou as ex:
+        print("Bucket already exists")
+    return bucket_name
+
+
+def _remove_aws_s3_bucket_content(bucket_name: str) -> None:
```
Comment (tkilias) [inline test/notebooks/nbtest_sagemaker.py:27]: we probably state this was copied from the sagemaker extension

File: test/notebooks/nbtest_sagemaker.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            RoleName=role_name,
+            AssumeRolePolicyDocument=assume_policy_document,
+            Description='This role is used for the CI Tests of the exasol.sagemaker-extension',
+        )
+    except iam_client.exceptions.EntityAlreadyExistsException as ex:
+        print(f"Role '{role_name}' already exists")
+    return role_name
+
+
+def _create_sagemaker_policy(iam_client) -> str:
```
Comment (tkilias) [inline test/notebooks/nbtest_sagemaker.py:60]: we probably state this was copied from the sagemaker extension

File: test/notebooks/nbtest_sagemaker.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    conf = Secrets(store_path, store_password)
+    conf.connection()
+    conf.save(CKey.aws_access_key_id, os.environ["NBTEST_AWS_ACCESS_KEY_ID"])
+    conf.save(CKey.aws_secret_access_key, os.environ["NBTEST_AWS_SECRET_ACCESS_KEY"])
+    conf.save(CKey.aws_region, os.environ["NBTEST_AWS_DEFAULT_REGION"])
+    conf.save(CKey.aws_bucket, bucket_name)
+    conf.save(CKey.aws_role, role_name)
+
+
+@pytest.fixture(scope="session")
```
Comment (tkilias) [inline test/notebooks/nbtest_sagemaker.py:110]: we probably state this was copied from the sagemaker extension

File: test/notebooks/nbtest_sagemaker.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+def aws_s3_bucket() -> str:
+    bucket_name = _create_aws_s3_bucket()
+    try:
+        yield bucket_name
+    finally:
+        _remove_aws_s3_bucket_content(bucket_name)
+
+
+@pytest.fixture(scope="session")
+def aws_sagemaker_role() -> str:
```
Comment (tkilias) [inline test/notebooks/nbtest_sagemaker.py:120]: we probably state this was copied from the sagemaker extension

**Review by tkilias — CHANGES_REQUESTED** [review]

File: test/notebooks/notebook_test_utils.py (hunk 1 above)
Comment (ahsimb) [reply test/notebooks/notebook_test_utils.py:106]: It should be handled by the pytest's tmp_path fixture.

**Review by tkilias — DISMISSED** [review]

**Review by tkilias — DISMISSED** [review]

**Review by tkilias — DISMISSED** [review]

**Review by tkilias — APPROVED** [review]

**Review by tkilias — DISMISSED** [review]

**Review by tkilias — APPROVED** [review]

**Pull request merged**